pkg/ruler/ruler.go

```diff
@@ -283,18 +333,22 @@ func NewRuler(cfg Config, manager MultiTenantManager, reg prometheus.Registerer,
 
 func newRuler(cfg Config, manager MultiTenantManager, reg prometheus.Registerer, logger log.Logger, directStore, cachedStore rulestore.RuleStore, limits RulesLimits, clientPool ClientsPool) (*Ruler, error) {
 	ruler := &Ruler{
-		cfg:            cfg,
-		directStore:    directStore,
-		cachedStore:    cachedStore,
-		manager:        manager,
-		registry:       reg,
-		logger:         logger,
-		limits:         limits,
-		clientsPool:    clientPool,
-		allowedTenants: util.NewAllowedTenants(cfg.EnabledTenants, cfg.DisabledTenants),
-		metrics:        newRulerMetrics(reg),
+		cfg:               cfg,
+		directStore:       directStore,
+		cachedStore:       cachedStore,
+		manager:           manager,
+		registry:          reg,
+		logger:            logger,
+		limits:            limits,
+		clientsPool:       clientPool,
+		outboundSyncQueue: newRulerSyncQueue(cfg.syncQueuePollFrequency()),
+		inboundSyncQueue:  newRulerSyncQueue(cfg.syncQueuePollFrequency()),
+		allowedTenants:    util.NewAllowedTenants(cfg.EnabledTenants, cfg.DisabledTenants),
+		metrics:           newRulerMetrics(reg),
 	}
 
+	ruler.outboundSyncQueueProcessor = newRulerSyncQueueProcessor(ruler.outboundSyncQueue, ruler.notifySyncRules)
```

Comment:
I didn't want to do it in `run()` to not block rules synching in case of any slowdown talking to other rulers. I would feel more comfortable if it's completely decoupled.